testplan/runners/pools/remote.py

```diff
@@ -704,19 +716,54 @@ def _add_workers(self):
             self._workers.add(worker, uid=instance['host'])
 
     def _start_workers(self):
-        num_workers = len(self._workers)
+        """Start all workers of the pool"""
+        for worker in self._workers:
+            self._conn.register(worker)
+        if self.pool:
+            self._workers.start_in_pool(self.pool)
+        else:
+            self._workers.start()
+
+    def _stop_workers(self):
+        if self.pool:
+            self._workers.stop_in_pool(self.pool)
+        else:
+            self._workers.stop()
+        # self._workers.stop()
+
+    def _start_thread_pool(self):
+        size = self.cfg.size
         try:
-            if num_workers > 2:
-                self.pool = ThreadPool(5 if num_workers > 5 else num_workers)
+            if size > 2:
+                self.pool = ThreadPool(5 if size > 5 else size)
         except Exception as exc:
             if isinstance(exc, AttributeError):
                 self.logger.warning(
                     'Please upgrade to the suggested python interpreter.')
 
-        super(RemotePool, self)._start_workers()
+    def starting(self):
+        self._start_thread_pool()
+        super(RemotePool, self).starting()
 
     def stopping(self):
+
+        sleeper = get_sleeper(1, 300)
+        while next(sleeper):
+            if any([worker.status.tag == worker.status.STARTING
+                   for worker in self._workers]):
+                self.logger.info('Waiting for workers to quit starting')
+            else:
+                break
+        else:
+            self.logger.warning(
```

Comment:
again, I think we should raise an exception and/or call self.abort() here - we don't expect to stop while the workers are still starting?

Reply:
The problem is starting and stopping could overlap. Some treadmill worker could come up and starting executing tests and even finished all tests while other workers are still initializing. One worker could also be restarting when testplan determines all tests are finished and it's time to stop.

Reply:
As above, let's try to prevent start/stop overlap.

Simple FSM: STARTING -> STARTED -> STOPPING -> STOPPED :)

Reply:
this is just best effort, due to treadmill nature, late joiner worker or restarted worker could still be starting while pool's stop is called. So the logic is kept here, see internal PR for treadmill pool change.